examples/gno.land/r/gov/dao/v3/memberstore/types.gno

```diff
@@ -0,0 +1,172 @@
+package memberstore
+
+import (
+	"std"
+
+	"gno.land/p/demo/avl"
+)
+
+type ErrMemberAlreadyExists struct {
+	Tier string
+}
+
+func (e *ErrMemberAlreadyExists) Error() string {
+	return "member already exists on tier " + e.Tier
+}
+
+type Member struct {
+	InvitationPoints int
+}
+
+func (m *Member) RemoveInvitationPoint() {
+	if m.InvitationPoints <= 0 {
+		panic("not enough invitation points")
+	}
+
+	m.InvitationPoints = m.InvitationPoints - 1
+}
+
+// MembersByTier contains all `Member`s indexed by their Address.
+type MembersByTier struct {
+	*avl.Tree // tier name -> address -> member
+}
+
+func NewMembersByTier() MembersByTier {
+	return MembersByTier{avl.NewTree()}
+}
+
+// GetTierSize tries to get how many members are on the specified tier. If the tier does not exists, it returns 0.
+func (mbt MembersByTier) GetTierSize(tn string) int {
+	tv, ok := mbt.Get(tn)
+	if !ok {
+		return 0
+	}
+
+	tree, ok := tv.(*avl.Tree)
+	if !ok {
+		return 0
+	}
+
+	return tree.Size()
+}
+
+// SetMember adds a new member to the specified tier. The tier index is created on the fly if it does not exists.
+func (mbt MembersByTier) SetMember(tier string, addr std.Address, member *Member) error {
+	_, t := mbt.GetMember(addr)
+	if t != "" {
+		return &ErrMemberAlreadyExists{Tier: t}
+	}
+
+	if ok := mbt.Has(tier); !ok {
+		mbt.Set(tier, avl.NewTree())
+	}
+
+	ms, _ := mbt.Get(tier)
+	mst := ms.(*avl.Tree)
+
+	mst.Set(string(addr), member)
+
+	return nil
+}
+
+// GetMember iterate over all tiers to try to find a member by its address. The tier ID is also returned if the Member is found.
+func (mbt MembersByTier) GetMember(addr std.Address) (m *Member, t string) {
+	mbt.Iterate("", "", func(tn string, msv interface{}) bool {
+		mst, ok := msv.(*avl.Tree)
+		if !ok {
+			panic("MembersByTier values can only be avl.Tree")
+		}
+
+		return mst.Iterate("", "", func(ma string, msv interface{}) bool {
+			if ma != string(addr) {
+				return false
+			}
+
+			mst, ok := msv.(*Member)
+			if !ok {
+				panic("MembersByTier values can only be *Member")
+			}
+
+			m = mst
+			t = tn
+
+			return true
+		})
+	})
+
+	return
+}
+
+// RemoveMember removes a member from any tier
+func (mbt MembersByTier) RemoveMember(addr std.Address) (t string) {
+	mbt.Iterate("", "", func(tn string, msv interface{}) bool {
+		mst, ok := msv.(*avl.Tree)
+		if !ok {
+			panic("MembersByTier values can only be avl.Tree")
+		}
+
+		_, removed := mst.Remove(string(addr))
+		t = tn
+		return removed
+	})
+
+	return
+}
+
+// GetTotalPower obtains the total voting power from all the specified tiers.
+func (mbt MembersByTier) GetTotalPower() float64 {
+	var out float64
```

Comment:
Could this be computed when the value changes and obtained directly without having to iterate?